src/menelaus/concept_drift/md3.py

```diff
@@ -0,0 +1,358 @@
+            y_train, y_test = self.reference_batch_target.iloc[train_index], self.reference_batch_target.iloc[test_index]
+            
+            duplicate_classifier.fit(X_train, y_train.values.ravel())
+
+            # record margin inclusion signals for all samples in this test band
+            signal_func_values = []
+            for i in range(len(X_test)):
+                sample_np_array = X_test.iloc[i].to_numpy()
+                margin_inclusion_signal = self.calculate_margin_inclusion_signal_other_classifier(sample_np_array, duplicate_classifier)
+                signal_func_values.append(margin_inclusion_signal)
+
+            # record margin density over this test band
+            margin_density = sum(signal_func_values) / len(signal_func_values)
+            margin_densities.append(margin_density)
+
+            # record accuracy of prediction over this test band
+            y_pred = duplicate_classifier.predict(X_test)
+            accuracy = accuracy_score(y_test, y_pred)
+            accuracies.append(accuracy)
+
+        # record performance - margin density
+        md = np.mean(margin_densities)
+        md_std = np.std(margin_densities)
+
+        # record performance - accuracy
+        acc = np.mean(accuracies)
+        acc_std = np.std(accuracies)
+
+        # return reference distribution statistics
+        return {
+            "len": len(data),
+            "md": md,
+            "md_std": md_std,
+            "acc": acc,
+            "acc_std": acc_std
+        }
+
+    def calculate_margin_density(self, data):
+        """
+        Calculate the total margin density of the batch of data passed in.
+
+        Args:
+            data (DataFrame): batch of data to calculate margin density for
+        """
+
+        signal_func = 0
+        for i in range(len(data)):
+            sample_np_array = data.iloc[i].to_numpy()
+            margin_inclusion_signal = self.calculate_margin_inclusion_signal(sample_np_array)
+            signal_func += margin_inclusion_signal
+        
+        return signal_func / len(data)
+
+    def update(self, new_sample):
+        """
+        Update the detector with a new sample.
+
+        Args:
+            new_sample (DataFrame): feature values/sample data for the new incoming sample
+        """
+        
+        if self.waiting_for_oracle == True:
+            raise ValueError(
+                """give_oracle_labels method must be called to provide detector with
+                labeled samples to confirm or rule out drift."""
+            )
+
+        if len(new_sample) != 1:
+            raise ValueError(
+                """This method is only available for data inputs in the form of 
+                a Pandas DataFrame with exactly 1 record."""
+            )
+            
+        if self.drift_state == "drift":
+            self.reset()
+            
+        super().update()
+
+        sample_np_array = new_sample.iloc[0].to_numpy()
```

Comment:
This is inconsistent with the other detectors, I think. You're already doing the work of picking out a single row at a time in your example loop, so I'm guessing this is because `X_test` is passed in as a new dataframe there, instead of via e.g. `X_test.iloc[i]` from the original dataframe.

Reply:
Ah, nah, not the `to_numpy` call, I mean the `new_sample.iloc[0]` -- I think the other detectors are set up such that the `iloc` is called like `det.update(X.iloc[0])`, so that call wouldn't be necessary.